### PR TITLE
Fix duplicate field document parsing exception in synthetic source

### PR DESCRIPTION
**Pull request by lkts**

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/XContentDataHelper.java [hunk 1]
```diff
@@ -235,9 +218,17 @@ private static DocumentParserContext cloneDocumentParserContext(
...
-        DocumentParserContext subcontext = context.switchParser(
-            XContentHelper.createParserNotCompressed(configuration, BytesReference.bytes(builder), context.parser().contentType())
+        XContentParser newParser = XContentHelper.createParserNotCompressed(
+            configuration,
+            BytesReference.bytes(builder),
+            context.parser().contentType()
         );
+        if (context.isDotExpandingEnabled()) {
+            // if we did dot expanding originally we need to continue to do so when we replace the parser
+            newParser = DotExpandingXContentParser.expandDots(newParser, context.path());
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/XContentDataHelper.java:228]: Can we just use `context.path()` here? Or do we need to create a new one? The path no longer align with fields new parser, I think?

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java [hunk 2]
```diff
@@ -352,7 +358,15 @@ public final DocumentParserContext addIgnoredFieldFromContext(IgnoredSourceField
             } else {
                 assert ignoredFieldWithNoSource != null;
                 assert ignoredFieldWithNoSource.value() == null;
+
+                // Don't expand dots when we use the parser to get part of the source
+                // since dot expanding can modify a document in a way that makes it invalid.
+                // See #encodeFlattenedToken().
+                boolean old = path().isWithinLeafObject();
+                path().setWithinLeafObject(true);
                 Tuple<DocumentParserContext, XContentBuilder> tuple = XContentDataHelper.cloneSubContext(this);
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:367]: Would just unwrapping the parser in `cloneSubContextParserConfiguration(...)` be possible and if so easier?

**Review by martijnvg — COMMENTED** [review]
Thanks for working on this! I left two questions.

File: server/src/main/java/org/elasticsearch/index/mapper/XContentDataHelper.java (hunk 1 above)
Comment (lkts) [reply server/src/main/java/org/elasticsearch/index/mapper/XContentDataHelper.java:228]: It should be the same, this is the parser that we return for "regular" indexing so it should be at the same point as before the clone.

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java (hunk 2 above)
Comment (lkts) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:367]: Yes, i think it's cleaner.

Comment (lkts) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java:356]: I tried using your approach with `getWrappedParser()` but `DotExpandingXContentParser` can wrap more than one parser and i am not sure what's going on at that time.

**Comment by lkts** [comment]
Closing in favor of #115822.

**Pull request closed without merging**